modules/fedimint-ln-server/src/lib.rs

```diff
@@ -1173,6 +1208,65 @@ impl Lightning {
             dbtx.remove_entry(&key).await;
         }
     }
+
+    /// Returns the challenge to the gateway that must be signed by the
+    /// gateway's private key in order for the gateway registration record
+    /// to be removed. The challenge is the concatenation of the gateway's
+    /// public key and the `valid_until` bytes. This ensures that the
+    /// challenges changes every time the gateway is re-registered and ensures
+    /// that the challenge is unique per-gateway.
+    async fn get_gateway_remove_challenge(
+        &self,
+        gateway_id: PublicKey,
+        dbtx: &mut DatabaseTransaction<'_>,
+    ) -> Option<sha256::Hash> {
+        if let Some(gateway) = dbtx.get_value(&LightningGatewayKey(gateway_id)).await {
+            let mut valid_until_bytes = gateway.valid_until.to_bytes();
+            let mut challenge_bytes = gateway_id.to_bytes();
+            challenge_bytes.append(&mut valid_until_bytes);
+            Some(sha256::Hash::hash(&challenge_bytes))
+        } else {
+            None
+        }
+    }
+
+    /// Removes the gateway registration record. First the signature provided by
+    /// the gateway is verified by checking if the gateway's challenge has
+    /// been signed by the gateway's private key.
+    async fn remove_gateway(
+        &self,
+        remove_gateway_request: RemoveGatewayRequest,
+        dbtx: &mut DatabaseTransaction<'_>,
+    ) -> anyhow::Result<()> {
+        let gateway_id = remove_gateway_request.gateway_id;
+        let our_peer_id = self.our_peer_id;
+        let signature = remove_gateway_request.signatures.get(&our_peer_id);
+
+        if signature.is_none() {
+            warn!("No signature provided for gateway: {gateway_id}");
+            return Err(anyhow::anyhow!(
+                "No signature provided for gateway {gateway_id}"
+            ));
+        }
```

Comment:
This could be an [`.ok_or_else(|| {…})`](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or_else):

```rust
let signature = remove_gateway_request.signatures.get(&our_peer_id)
    .ok_or_else(|| {
        warn!("No signature provided for gateway: {gateway_id}");
        anyhow::anyhow!(
            "No signature provided for gateway {gateway_id}"
        );
    })?;
```